### DISCUSSION TITLE
Retry timin clarification  for client.connect, client.read_holding_register, etc (ModbusClient.AsyncModbusTcpClient pymodbus v3.68)

### DISCUSSION THREAD

**Discussion by Ducati748ss** [pymodbus-dev/pymodbus#2171] [Help] [answered]
Hi
Is my following understanding of the retry timings for client.connect client.read_holding_register, etc (ModbusClient.AsyncModbusTcpClient  pymodbus v3.68) correct?

As my application must rapidly provide a control system setpoint, I need to make sure I understand exactly the retry process and the exact timings I can expect.

1)	The “reconnect_delay” doubles on each retry?
2)	Once the “reconnect_delay_max” is reached, any future retries will be for its duration?
3)	The  “timeout” value will always take precedence, even if the number of retries (“retries”) is not met? 
4)	There is an initial try to communicate then N retries, as set by the value of “retries”?

Therefore, with these values:
      **timeout = 5s
      retries = 3
      reconnect_delay = 0.4s
      reconnect_delay_max = 1.0s**

Then approximate timings for the waits would be:
      **Try 0    = 0.4s
      Retry 1 = 0.4s
      Retry 2 = 0.8s
      Retry 3 = 1.0s 		          (capped at “reconnect_delay_max”)
      wait 5.0s - 2.6s = 2.4s
      return** 

That is, the total time trying the 4 tries is approximately 2.6 seconds, so now wait 2.4 seconds for a response from Retry 3 until the “timeout” of 5 seconds expires, then return.

Thank you very much
Mark Norman

**Comment by janiversen** [comment, marked as answer]
Sound just about right.
